### PR TITLE
 refactoring AuthorizationHandlerManager

### PR DESCRIPTION
**Pull request by nuwandiw**
### Proposed changes in this pull request
- Refactoring AuthorizationHandlerManager class. Removing the duplicate grant type validation and improving readability.

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 1]
```diff
@@ -75,53 +75,45 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
-        if (!responseHandlers.containsKey(responseType)) {
-            log.warn("Unsupported Response Type : " + responseType +
-                    " provided  for user : " + authzReqDTO.getUser());
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNSUPPORTED_RESPONSE_TYPE,
-                    "Unsupported Response Type!");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        OAuthAuthzReqMessageContext authzReqMsgCtx = getoAuthAuthzReqMessageContext(authzReqDTO);
+        ResponseTypeHandler authzHandler = responseHandlers.get(authzReqDTO.getResponseType());
+        OAuth2AuthorizeRespDTO authorizeRespDTO = validateAuthzRequest(authzReqDTO, authzReqMsgCtx, authzHandler);
+        if (foundErrorResponse(authorizeRespDTO)) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Better to have a debug log

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 2]
```diff
@@ -75,53 +75,45 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
-        if (!isAuthorizedClient) {
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT,
-                    "The authenticated client is not authorized to use this authorization grant type");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+    private OAuth2AuthorizeRespDTO validateAuthzRequest(OAuth2AuthorizeReqDTO authzReqDTO,
+                                                        OAuthAuthzReqMessageContext authzReqMsgCtx,
+                                                        ResponseTypeHandler authzHandler)
+            throws IdentityOAuth2Exception {
+        OAuth2AuthorizeRespDTO authorizeRespDTO = new OAuth2AuthorizeRespDTO();
+        if(!validateResponseType(authzReqDTO, authorizeRespDTO)) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Better to have a debug log

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 3]
```diff
@@ -75,53 +75,45 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
             return authorizeRespDTO;
         }
-
-        boolean accessDelegationAuthzStatus = authzHandler.validateAccessDelegation(authzReqMsgCtx);
-        if(authzReqMsgCtx.getProperty("ErrorCode") != null){
-            authorizeRespDTO.setErrorCode((String)authzReqMsgCtx.getProperty("ErrorCode"));
-            authorizeRespDTO.setErrorMsg((String)authzReqMsgCtx.getProperty("ErrorMsg"));
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        if (!validateClient(authzReqDTO, authorizeRespDTO, authzReqMsgCtx, authzHandler)) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Better to have a debug log

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 4]
```diff
@@ -75,53 +75,45 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
             return authorizeRespDTO;
-        } else if (!accessDelegationAuthzStatus) {
-            log.warn("User : " + authzReqDTO.getUser() +
-                    " doesn't have necessary rights to grant access to the resource(s) " +
-                    OAuth2Util.buildScopeString(authzReqDTO.getScopes()));
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT,
-                    "Authorization Failure!");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        }
+        if (!validateAccessDelegation(authzReqDTO, authorizeRespDTO, authzReqMsgCtx, authzHandler)) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Better to have a debug log

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 5]
```diff
@@ -130,29 +122,74 @@ public OAuth2AuthorizeRespDTO handleAuthorization(OAuth2AuthorizeReqDTO authzReq
...
-	    OAuth2Util.clearAuthzRequestContext();
-	}
-	
-        return authorizeRespDTO;
+    private boolean validateAccessDelegation(OAuth2AuthorizeReqDTO authzReqDTO, OAuth2AuthorizeRespDTO authorizeRespDTO,
+                                             OAuthAuthzReqMessageContext authzReqMsgCtx,
+                                             ResponseTypeHandler authzHandler) throws IdentityOAuth2Exception {
+        boolean accessDelegationAuthzStatus = authzHandler.validateAccessDelegation(authzReqMsgCtx);
+        if (!accessDelegationAuthzStatus) {
+            log.warn("User : " + authzReqDTO.getUser() +
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Would DEBUG level better than WARN as this is expected business case.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 6]
```diff
@@ -130,29 +122,74 @@ public OAuth2AuthorizeRespDTO handleAuthorization(OAuth2AuthorizeReqDTO authzReq
...
+        }
+        return true;
+    }
+
+    private boolean validateClient(OAuth2AuthorizeReqDTO authzReqDTO, OAuth2AuthorizeRespDTO authorizeRespDTO,
+                                   OAuthAuthzReqMessageContext authzReqMsgCtx, ResponseTypeHandler authzHandler)
+            throws IdentityOAuth2Exception {
+        boolean isAuthorizedClient = authzHandler.isAuthorizedClient(authzReqMsgCtx);
+        if (!isAuthorizedClient) {
+            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT,
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Better to have a debug log

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java (hunk 1 above)
Comment (nuwandiw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Done

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java (hunk 2 above)
Comment (nuwandiw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Added debug logs within validateResponseType() method

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java (hunk 3 above)
Comment (nuwandiw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Added debug logs within validateClient() method

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java (hunk 4 above)
Comment (nuwandiw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Added debug logs within validateAccessDelegation() method

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java (hunk 6 above)
Comment (nuwandiw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Done

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 7]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
-        if (!responseHandlers.containsKey(responseType)) {
-            log.warn("Unsupported Response Type : " + responseType +
-                    " provided  for user : " + authzReqDTO.getUser());
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNSUPPORTED_RESPONSE_TYPE,
-                    "Unsupported Response Type!");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        OAuthAuthzReqMessageContext authzReqMsgCtx = getoAuthAuthzReqMessageContext(authzReqDTO);
+        ResponseTypeHandler authzHandler = responseHandlers.get(authzReqDTO.getResponseType());
+        OAuth2AuthorizeRespDTO authorizeRespDTO = validateAuthzRequest(authzReqDTO, authzReqMsgCtx, authzHandler);
+        if (foundErrorResponse(authorizeRespDTO)) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: shall we rename foundErrorResponse ---> isErrorResponseFound

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 8]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
+            if (log.isDebugEnabled()) {
+                log.debug("Client validation failed for user : " + authzReqDTO.getUser() +
+                        ", for client : " + authzReqDTO.getConsumerKey());
+            }
+            return false;
+        }
+        return true;
+    }
+
+    private OAuthAuthzReqMessageContext getoAuthAuthzReqMessageContext(OAuth2AuthorizeReqDTO authzReqDTO)
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: getoAuthAuthzReqMessageContext -> get**O**AuthAuthzReqMessageContext

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 9]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
+            throws IdentityOAuth2Exception, InvalidOAuthClientException {
+        OAuthAuthzReqMessageContext authzReqMsgCtx;
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Merge two lines.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 10]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
+        authzReqMsgCtx = new OAuthAuthzReqMessageContext(authzReqDTO);
+        // loading the stored application data
+        OAuthAppDO oAuthAppDO = getAppInformation(authzReqDTO);
+        authzReqMsgCtx.addProperty("OAuthAppDO", oAuthAppDO);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Use a constant

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 11]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
+            }
+            return false;
+        }
+        return true;
+    }
+
+    private OAuthAuthzReqMessageContext getoAuthAuthzReqMessageContext(OAuth2AuthorizeReqDTO authzReqDTO)
+            throws IdentityOAuth2Exception, InvalidOAuthClientException {
+        OAuthAuthzReqMessageContext authzReqMsgCtx;
+        authzReqMsgCtx = new OAuthAuthzReqMessageContext(authzReqDTO);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Shall we change the authzReqMsgCtx variable name as -> authorizeRequestMessageContext to improve readability

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 12]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
-        OAuth2AuthorizeRespDTO authorizeRespDTO = new OAuth2AuthorizeRespDTO();
-
-        if (!responseHandlers.containsKey(responseType)) {
-            log.warn("Unsupported Response Type : " + responseType +
-                    " provided  for user : " + authzReqDTO.getUser());
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNSUPPORTED_RESPONSE_TYPE,
-                    "Unsupported Response Type!");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        OAuthAuthzReqMessageContext authzReqMsgCtx = getoAuthAuthzReqMessageContext(authzReqDTO);
+        ResponseTypeHandler authzHandler = responseHandlers.get(authzReqDTO.getResponseType());
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Move to a private method -> getResponseHandler

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 13]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
+                                  OAuthAuthzReqMessageContext authzReqMsgCtx, ResponseTypeHandler authzHandler)
+            throws IdentityOAuth2Exception {
         boolean scopeValidationStatus = authzHandler.validateScope(authzReqMsgCtx);
         if (!scopeValidationStatus) {
-            log.warn("Scope validation failed for user : "
-                    + authzReqDTO.getUser() + ", for the scope : "
-                    + OAuth2Util.buildScopeString(authzReqDTO.getScopes()));
-            handleErrorRequest(authorizeRespDTO,
-                    OAuthError.CodeResponse.INVALID_SCOPE, "Invalid Scope!");
+            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.INVALID_SCOPE, "Invalid Scope!");
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Shall we use static import for  OAuthError

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 14]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
+
+    private boolean validateResponseType(OAuth2AuthorizeReqDTO authzReqDTO, OAuth2AuthorizeRespDTO authorizeRespDTO) {
+        if (!responseHandlers.containsKey(authzReqDTO.getResponseType())) {
+            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNSUPPORTED_RESPONSE_TYPE,
+                    "Unsupported Response Type!");
+            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+            if (log.isDebugEnabled()) {
+                log.debug("Unsupported Response Type : " + authzReqDTO.getResponseType() +
+                        " provided for user : " + authzReqDTO.getUser() +
+                        ", for client " + authzReqDTO.getConsumerKey());
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: for client ->  "for client : "

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 15]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
-        if (!isAuthorizedClient) {
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT,
-                    "The authenticated client is not authorized to use this authorization grant type");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+    private OAuth2AuthorizeRespDTO validateAuthzRequest(OAuth2AuthorizeReqDTO authzReqDTO,
+                                                        OAuthAuthzReqMessageContext authzReqMsgCtx,
+                                                        ResponseTypeHandler authzHandler)
+            throws IdentityOAuth2Exception {
+        OAuth2AuthorizeRespDTO authorizeRespDTO = new OAuth2AuthorizeRespDTO();
+        if(!validateResponseType(authzReqDTO, authorizeRespDTO)) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: shall we move to a private method : isInvalidResponseType

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 16]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
+            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.INVALID_SCOPE, "Invalid Scope!");
             authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
-            return authorizeRespDTO;
-        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Scope validation failed for user : " + authzReqDTO.getUser() + ", for the scope(s) : "
+                        + OAuth2Util.buildScopeString(authzReqDTO.getScopes()));
+            }
+            return false;
+        } else if (authzReqMsgCtx.getApprovedScope() == null || authzReqMsgCtx.getApprovedScope().length == 0) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Shall we move to a private method

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 17]
```diff
@@ -75,84 +75,135 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
...
+                log.debug("Error response received for authorization request by user : " + authzReqDTO.getUser()  +
+                        ", client : " + authzReqDTO.getConsumerKey() + ", scope : " +
+                        OAuth2Util.buildScopeString(authzReqDTO.getScopes()));
+            }
             return authorizeRespDTO;
         }
+        try {
+	    // set the authorization request context to be used by downstream handlers. This is introduced as a fix for
+	    // IDENTITY-4111
+	    OAuth2Util.setAuthzRequestContext(authzReqMsgCtx);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: Couldn't find any place which uses this threadLocal variable. Shall we more analyze this is required or not

Comment (nuwandiw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: We give the support to extend the org.apache.oltu.oauth2.as.issuer.OAuthIssuer implementation in OauthTokenIssuerImpl (check [1]). org.apache.oltu.oauth2.as.issuer.OAuthIssuer APIs do not accept any messageContext data. authzHandler.issue() method may call the implemented OAuthIssuer APIs in downstream. In that case, OAuthIssuer APIs need the threadLocal to retrieve authzReqMsgCtx data. Therefore cannot remove this.

[1] https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OauthTokenIssuerImpl.java#L29

**Pull request merged**